### DISCUSSION TITLE
Add watermark to image files with different sizes from command line

### DISCUSSION THREAD

**Discussion by Elfangor93** [ImageMagick/ImageMagick#3043] [Help]
In the image gallery tool JoomGallery (written im php), we are using ImageMagick for processing the images (resizing images, rotating images,...).
I am currently working on a way to watermark images using ImageMagick. I therefore assemble an imagick command using php and execute the command using the [@exec()](https://www.php.net/manual/en/function.exec) function.
I would like to resize a given png watermark file and copy it with a specific opacity at a specific position on a given jpg file.
Currently I am using the following command:

`convert "image.jpg" ( "watermark.png" -resize "1200x600" ) -gravity "northwest" -geometry "+0+100" -define compose:args=75,100 -compose dissolve -composite "image.jpg"`

where the resize dimension (1200x600) and the position of the watermark (+0+100) gets calculated from php depending on the configurations in the Gallery.

This command works in most cases, but on some systems this command throws an error and will not be executed. I guess this command has some compatibility issues on some systems...
Is there a better way to assemble the imagick command?
I am searching for a command that is working on as much systems as possible (Linux or Windows and ImageMagick v6.x or v7.x).

Tanks in advance for your help...

**Comment by fmw42** [comment]
I do not see anything wrong with the command. What version of IM and what platform does it fail? Please provide example images that fail. If using IM 7, use magick not convert.

**Comment by Elfangor93** [comment]
platform information:
Linux kernel version 4.18.0-147.8.1.e18.lve.x86_64
IM information:
ImageMagick 6.9.4-10 Q16 x86_64 2019-12-01

If I run the command on this platform I get a syntax error:
![grafik](https://user-images.githubusercontent.com/39154009/103201372-0c42fe00-48f0-11eb-96b7-4014e217e5e9.png)

Without resizing of the watermark it works. It seems as if there is a problem with the brackets on some platforms...
`convert "image.jpg" "watermark.png" -gravity "northwest" -geometry "+0+100" -define compose:args=75,100 -compose dissolve -composite "image.jpg"`

**Reply by Elfangor93** [reply]
platform information:
XAMPP-Shell on Windows 10 (64-bit)
IM information:
ImageMagick 7.0.10-16 Q16 x64 2020-05-31

If I run the command on this platform everything works fine.

**Reply by Elfangor93** [reply]
platform information:
Debian GNU/Linux 10 (4.19.0-13-amd64)
IM information:
ImageMagick 6.910-23 Q16 x86_64 20190101`

syntax error near unexpected token `('

**Reply by Elfangor93** [reply]
platform information:
Debian GNU/Linux 10 (4.19.0-13-amd64)
IM information:
ImageMagick 7.0.10-53 Q16 x64_64 2020-12-28

syntax error near unexpected token `('

**Reply by Elfangor93** [reply]
platform information:
CMD-Shell on Windows 10 (64-bit)
IM information:
ImageMagick 7.0.10-16 Q16 x64 2020-05-31

If I run the command on this platform everything works fine.

**Reply by Elfangor93** [reply]
platform information:
PowerShell on Windows 10 (64-bit)
IM information:
ImageMagick 7.0.10-16 Q16 x64 2020-05-31

![grafik](https://user-images.githubusercontent.com/39154009/103216228-d0bb2a80-4915-11eb-8d42-dd995d8338c7.png)

**Comment by GeeMack** [comment]
If the parentheses are the only thing causing a problem, you can avoid using them in this instance by reading in the watermark image and resizing it first, then using "+swap" to get the images in the correct order for the rest of the command. Maybe start with something like this...

    convert "watermark.png" -resize "1200x600" "image.jpg" +swap ...

That reads in the watermark and resizes it, then the main image is read into the command, then the positions of those two images are swapped in the list, and the command continues as you've already written it.

**Reply by Elfangor93** [reply]
Thats it. Now it works on all platforms I am able to test on...
Thanks for your help!
